content/post/2017-05-04-firmware-update.md

```diff
@@ -0,0 +1,47 @@
+---
+title: Firmware Updates
+#subtitle:
+date: 2017-05-
+author: Justin Schneck
+draft: false
+tags: ["nerves", "fwup", "reactor", "bootloader"]
+---
+
+Firmware updates are an important aspect to all parts of the device lifecycle, not just production. In development, they can save you time not burning SD cards by using mechanisms like pushing firmware bundles over the network. Network updates can also pave the way to automate executing canary tests using real hardware. Firmware updates are important to these stages because they are the only way to modify the contents of the read only filesystem.
+
+# What is a Firmware Bundle?
+Firmware bundles are created using [fwup](https://github.com/fhunleth/fwup/). In Nerves, Firmware bundles are `.fw` archive files that are constructed as the output of calls to `mix firmware`. You can think of these bundles as instructions attached to a payload. In modern Nerves based Mix projects, you'll find these bundles at `_build/#{Mix.target}/#{Mix.Env}/nerves/images`, for example, if we were in a project called `my_app` for Raspberry Pi Zero and running in dev, `_build/rpi0/dev/nerves/images/my_app.fw`.
+
+The instructions contained in the bundle are declared in the fwup.conf included in the Nerves system. These instructions contain tasks like `complete` and `upgrade`.  
+
+When the `complete` task is applied, the destination device is written fresh, like an initial install producing a layout of 2 slots for firmware on the device. This A/B layout provides some safeguards by retaining a known good working firmware. For example, when an `upgrade` task is applied, the new firmware is written to the inactive slot on the device and made active on next the next boot.
+
+# How Do I Apply Them?
+To apply a firmware bundle you need `fwup` installed, the destination device (ex: `/dev/mmcblk0`), and a task to execute from the fwup.conf (ex: `upgrade`). Fwup is already included in all Nerves systems. you can interact with it from a console session running on the device.
+
```

Comment:
A bit strange to show invoking this from hex, rather than from the shell, but having the shell used a few lines later.